### DISCUSSION TITLE
Use-cases for type config decorator?

### DISCUSSION THREAD

**Discussion by vhenzl** [webonyx/graphql-php#921] [General]
What are the (intended) use-cases for type config decorator? There isn't much documentation. And issues usually mention it in relation to enums and custom scalars; nothing much there about resolvers.

My understanding is that the decorator is a remnant from `graphql-js` before [v0.12](https://github.com/graphql/graphql-js/releases/tag/v0.12.0). But even in the older versions, it never was intended for adding resolvers: https://github.com/graphql/graphql-js/blob/v0.11.7/src/utilities/buildASTSchema.js#L126-L127

The decorator was intentionally preserved in #248 and @vladar's comment (https://github.com/webonyx/graphql-php/pull/248#issuecomment-370201272) suggests it can be used for resolvers as well.

I have search through the code and issues and there aren't many examples. They usually show code with `resolveFields` (e.g. https://github.com/webonyx/graphql-php/issues/163#issuecomment-327013069), only one shows how to add individual field resolvers (https://github.com/webonyx/graphql-php/issues/179#issuecomment-334914656).

But one is more than nothing, so I guess code like this is fine and legal:

```php
$documentNode = Parser::parse('
    type Query {
        hello: String
    }
');

$typeConfigDecorator = static function (array $typeConfig) {
    if ($typeConfig['name'] === 'Query') {
        $fields = $typeConfig['fields'] ?? [];
        $typeConfig['fields'] = static function () use ($fields): array {
            if ($fields instanceof \Closure) {
                $fields = $fields();
            }
            $fields['hello']['resolve'] = static fn(): string => 'Ahoy!';
            return $fields;
        };
    }

    return $typeConfig;
};

$schema = BuildSchema::build($documentNode, $typeConfigDecorator);
```

The lack of examples makes me wonder. **Do people use it in this way at all? Are there any cons & pros comparing to `resolveFields` and `defaultFieldResolver`?** Or does everyone build their schema programmatically without SDL? (Personally, in one of my projects I use PHP objects, in the other with SDL I abuse the default resolver to add "middlewares".)

### Context
The context for this question and the reason why I went down this rabbit hole is the recent addition of `$typeConfigDecorator` to  `SchemaExtender::extend()`. I made a comment in the PR (https://github.com/webonyx/graphql-php/pull/871#issuecomment-905220352), but there is more to it.

Given that the above approach with setting individual field resolves is valid (and that I tested the current `extend()`'s behaviour correctly), there is a problem with resolvers for extended types.

```php
$documentNode1 = Parser::parse('
    type Query {
        hello: String
    }
');

$typeConfigDecorator1 = static function (array $typeConfig) {
    if ($typeConfig['name'] === 'Query') {
        $typeConfig['resolveField'] = static fn(): string => '*default 1*'; // 👀

        $fields = $typeConfig['fields'] ?? [];
        $typeConfig['fields'] = static function () use ($fields): array {
            if ($fields instanceof \Closure) {
                $fields = $fields();
            }
            $fields['hello']['resolve'] = static fn(): string => 'Hey!'; // 👀
            return $fields;
        };
    }

    return $typeConfig;
};

$schema = BuildSchema::build($documentNode1, $typeConfigDecorator1);

$documentNode2 = Parser::parse('
      extend type Query {
        bye: String
      }
');

$typeConfigDecorator2 = static function (array $typeConfig) {
    if ($typeConfig['name'] === 'Query') {
        $typeConfig['resolveField'] = static fn(): string => '*default 2*'; // 👀

        $fields = $typeConfig['fields'] ?? [];
        $typeConfig['fields'] = static function () use ($fields): array {
            if ($fields instanceof \Closure) {
                $fields = $fields();
            }
            $fields['bye']['resolve'] = static fn(): string => 'See ya!'; // 👀
            return $fields;
        };
    }

    return $typeConfig;
};

$extendedSchema = SchemaExtender::extend($schema, $documentNode2, [], $typeConfigDecorator2);

$query = '{ 
    hello
    bye
}';

$result = GraphQL::executeQuery($extendedSchema, $query);
```

You can't add a resolver for the field `bye` nor overwrite `resolveFields` – the second decorator doesn't run for `Query` type at all:

```diff
--- Expected
+++ Actual
@@ @@
-'{"data":{"hello":"Hey!","bye":"See ya!"}}'
+'{"data":{"hello":"Hey!","bye":"*default 1*"}}'
```

I don't know how much the implementation of `SchemaExtender` aligns with or deviates from the reference implementation, but we can't rely on it much anyway as this "type config decorator" is a `graphql-php` specific extension. So we have to decide for ourselves how it should work, what's allowed/supported and what isn't. 

I'm not sure if technically possible, but the "obvious" solution is to allow for `SchemaExtender` to apply the decorator not only to the new types from the extending document but also to existing types from schema. But then we have to ask how the configs should be merged. And for example, what should happen to `resolveField` (which is AFAIK also non-standard extension), because it can't be simply replaced or merged.

### Use-cases for type config decorator?
Back to my initial question. What are the use-cases for type config decorator?

- Adding enums & custom scalars. (Is there an alternative way to do that? How is it done in `graphql-js`?)
- Adding resolvers.
- Specifying types to support interfaces and unions (added 27/08/21, see https://github.com/webonyx/graphql-php/pull/871#issuecomment-905621746)
- Something else?

I'm not sure how about alternatives for custom types. But for resolvers, would it be feasible to port `makeExecutableSchema` from [`graphql-tools`](https://www.graphql-tools.com/docs/generate-schema)? And maybe drop the type config decorator altogether?

**Comment by spawnia** [comment]
Is this just a matter of overwriting `resolveField` through `SchemaExtender`? Or perhaps its precedence with individual field resolvers?

**Reply by vhenzl** [reply]
Both. And more.

Right now, it runs only for new types. It doesn't run for existing (i.e. extended) types at all, so you can't
- add field resolvers for the new fields
- nor add new `resolveField` to handle all the new field

Should be the extender allowed to modify existing types, then the questions are:
- If `resolveField` exists in both the existing schema and the decorator, 
	- should the latter replace the existing one (and take over its responsibilities)?
	- or should the latter somehow wrap the existing one and run it internally (before? after?) its own implementation?
- What should be logic generally? The precedence of individual field resolvers over `resolveField` is a good question.
	- Should individual field resolvers always take precedence regardless of how late they were specified and should `resolveField` serve as a fallback (as it does without extending)?
	- Should it be possible to overwrite already existing individual field resolvers?
	- How does the default resolver fit there?
	- What are the implications of having multiple type config decorators vs only one in the last `extend()`?

**Reply by spawnia** [reply]
I think a good starting point to answer those questions is to add a pull request with a failing test case. Use the API as you think it should work, then we can work backwards to achieve that dream API.

**Reply by vhenzl** [reply]
Yeah, that could help. 

How about my other questions? Mainly, the (other than providing resolvers) use-cases for type config decorator? Answers to them are important inputs to be able to craft the tests.

**Reply by spawnia** [reply]
I have not added the feature, nor have I ever used it. In my main project, we build the schema from SDL ourselves.

**Reply by vhenzl** [reply]
I'm aware that you haven't been a maintainer for that long. What's Vladimir's role nowadays, is he still involved with the project? He could have some insight…

**Comment by vhenzl** [comment]
I have done some research and here is what I've learnt.

There is a feature request (graphql/graphql-js#1858) for `graphql-js` that request basically the same that  `graphql-php`'s type config decorator provides. The issue sits there for some time now but is still opened and planned to be resolved.

In light of that, looking at solutions like `graphql-tools`'s `makeExecutableSchema` doesn't make much sense and we should focus on improving the type config decorator.

### The use-cases for type config decorator
The use-cases I gathered are:

- Support for custom scalar (`serialize`, `parseValue`, `parseLiteral`)
- Specifying values for enums
- Support for interfaces and unions (`resolveType` and `isTypeOf`)
- Adding resolvers (individual field `resolve` and type level `resolveField` functions)

It also matches with what's requested in graphql/graphql-js#1858.

In other words, the overall **use-case for type config decorator is to make the schema executable**. But that's pretty much all, the decorator shouldn't be used for things like modifying the schema (that's what AST and visitors are for).

### Type config decorator validation
A validation (sanity checks) should be introduced for the decorator to specify how it can manipulate the schema. These rules would apply to both `build()` and `extend()`. Then, on top of these general rules, we can specify rules for schema extending.

My opinion is that the decorator should be forbidden to touch config options coming from SDL/AST (`name`, `description`, `interfaces` of Object Type config, `name`, `type`, `args`, `description` of Field config, `name`, `types`, `description` of Union Type config, etc). The SDL document should be the owner of these data.

Currently, the decorator let you change anything, which in most cases results in an error when the decorator is applied or during execution.

I might miss something (comments welcome), but I reckon only these should be allowed:
- Object Type config: `isTypeOf` and `resolveField`
- Field config: `resolve`, `complexity`
- Scalar config: `serialize`, `parseValue`, `parseLiteral`
- Enum config: `values`
- Interface config: `resolveType`
- Union config: `resolveType`
- Any config: additional custom (i.e. unknown) fields as per https://webonyx.github.io/graphql-php/type-definitions/object-types/#custom-metadata

### Rules for schema extending
With just a limited set of options that can be used with decorators, it's easier to define rules for them.

And basically, everything boils down to a single question: what should happen if a config option for a particular type is specified multiple times (in `build()` and then again in `extend()`, or different `extend()` calls).

I don't think the schema should anyhow keep track of all values (by creating a stack of them, wrapping them to closure, or whatever the implementation could be). I reckon it should work on the last win principle - the later value simply overwrites the previous one.

This simple overwriting should work without problem for most of the options. However, question is, if an error shouldn't be thrown when a value for the particular option already exists.

There is no technical reason to do that, it's more a sanity check. I fail to see a valid use case when someone would set for example a field's `resolve` or object's `isTypeOf` during `build()` and then overwritten it immediately in a consequent `extend()`. It's most likely a mistake rather than an intention. If you have a real-world example, I'm eager to see it. 

Some other options (namely `resolveField` and `resolveType`) can always be overwritten without a warning. However, we would likely use the previous value to implement the new function. It can be done easily like this:

```php
$typeConfigDecorator = static function (array $typeConfig): array {
    if ($typeConfig['name'] === 'Query') {
        $resolveFieldFn = $typeConfig['resolveField'];
        $typeConfig['resolveField'] = static function ($source, $args, $context, $info) use ($resolveFieldFn) {
            // first handle the fields added by this extension
            if ($info->fieldName === 'aNewField') {
                return 'a new field value';
            }
            // then let the existing resolver handle the original fields
            return $resolveFieldFn($source, $args, $context, $info);
        };
    }
    return $typeConfig;
};
```
---
This is still a WIP, comments are welcome. I'll soon push a PR with tests to see all the above in code.
